[wasm] introduce typescript modules

pavelsavara opened the pull request:
# Goals
Introduce typescript and break up the code to multiple, better encapsulated modules.
- use EMSDK nodejs for the build
- use [rollup.js](https://rollupjs.org/guide/en/) to produce ES6 [IFFE](https://en.wikipedia.org/wiki/Immediately_invoked_function_expression) because emscripten doesn't know how to consume ES6 _modules_ on input.
- the ts output file `runtime.iffe.js` is compiled with emcc and AOT as [`--pre-js`](https://emscripten.org/docs/tools_reference/emcc.html)
- `package.json` references explicit versions of dev-time only dependencies. They are not necessary on link time. 
- dependencies are resolved with `npm ci` in CI pipeline to make sure we get exactly same version as in lock file. 
- we also run `npm audit` in CI so that we are sure we don't have new security issue in the CI. 
- we have small rollup plugin which checks sha256 hash of the output and doesn't overwrite it if it's the same. So that incremental build/link is not broken.
- `ignore_pdb_load_errors` is now `Module.config` option

# Non-goals
- just `library_mono.js` , the rest will be in follow up PRs
- functional changes
- later: lint, renames, more cleanup
- API change toward embedders like Blazor
- sourcemaps

kg commented on src/mono/wasm/runtime/rollup.config.js:
```diff
@@ -0,0 +1,22 @@
+import typescript from '@rollup/plugin-typescript';
+import { terser } from 'rollup-plugin-terser';
+
+export default defineConfig({
```
We should definitely put a detailed comment here or in another file explaining what rollup does for us based on this configuration

pavelsavara replied:
Emcc doesn't know how to load ES6 module, that's why we need the whole rollup.js
We only use it to squash all the ES6 modules typescript generated into single file wrapped as IFFE.
While we are at it, we could also minify the code in Release mode.

kg commented on src/mono/wasm/runtime/package.json:
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "dotnet-runtime",
```
Is it convention to have a package.json like this if this package is never meant to be uploaded to NPM?

pavelsavara replied:
It is a project file essentially, it drives tooling and install dependencies.
On uploading it to npm, I'm not sure if this package it the one we will publish, when nodeJS port is ready. I guess 50% chance.

kg commented on src/mono/wasm/runtime/src/mono/init.ts:
```diff
@@ -0,0 +1,533 @@
+	});
+}
+
+type Context = {
```
we should probably give this a more specific name

kg commented on src/mono/wasm/runtime/src/mono/scheduling.ts:
```diff
@@ -0,0 +1,78 @@
+    }
+}
+
+function pump_message() {
```
I think there are cases where we'll want to call this manually, so we may want to export it

pavelsavara replied:
I believe that `schedule_background_exec` was/is exactly that.

kg replied:
```schedule_background_exec``` queues a pump operation, you might want to synchronously call pump_message to implement something like ```Task.Wait```

kg commented on src/mono/wasm/runtime/tsconfig.json:
```diff
@@ -0,0 +1,20 @@
+    "noImplicitAny": true,
+    "noEmitOnError": true,
+    "removeComments": false,
+    "sourceMap": false,
```
Is there a reason why you turned off source maps?

pavelsavara replied:
There are 4 steps which map has to survive in sync with the code positions.
1) typescript transpilation
2) rollup-plugin-terser
3) emcc (closure)
4) AOT and customer AOT

I thought I will leave it for later as it seems like rabbit hole.

am11 commented:
Would it be possible to make `WasmEnableES6` default to `true` with this change? https://github.com/dotnet/runtime/blob/4d411739ce1d9db68dfe12227663479c9d71c50d/src/mono/wasm/wasm.proj#L10
`WasmEnableES6` results in importable ES6 module for dotnet in deno environment. It's also compatible in browser (since WASM and ES6 are mutually supported everywhere).

pavelsavara commented on src/mono/wasm/runtime/src/mono/init.ts:
```diff
@@ -0,0 +1,533 @@
+import { toBase64StringImpl } from './base64';
+import { AOTProfilerOptions, CoverageProfilerOptions, mono_wasm_init_aot_profiler, mono_wasm_init_coverage_profiler } from './profiler';
+
+const mono_wasm_ignore_pdb_load_errors = true;
```
made startup argument out of it

pavelsavara commented:
> Would this make it possible to make `WasmEnableES6` default to `true` with this change?
> 
> https://github.com/dotnet/runtime/blob/4d411739ce1d9db68dfe12227663479c9d71c50d/src/mono/wasm/wasm.proj#L10
> 
> 
> `WasmEnableES6` results in importable ES6 module for dotnet in deno environment. It's also compatible in browser (since WASM and ES6 are mutually supported everywhere).

I would love to, but I need to work thru all the consequences. Our unit test infrastructure and also the way how external projects like ASP.NET consume us. So most likely it would not be in this PR.

But it's on my wish list too.

pavelsavara committed:
- refactor wasm javascript as typescript
- use nodeJs from EMSDK
- use rollup.js for packing code as ES6 IFFE for emcc consumption
- incremental build Co-authored-by: Radek Doulik <radekdoulik@gmail.com>
- types Co-authored-by: Daniel Genkin <genkind@gmail.com>

pavelsavara commented:
OSX failure is unrelated timeout.

pavelsavara committed:
move mono_wasm_load_bytes_into_heap

pavelsavara commented:
@radical I introduced new file to AOT inputs. I don't know how it would impact VS/workloads. We will probably remove or rename all the other `.js` files too in next PR. Would something break in the meantime ?

radical commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -217,7 +217,8 @@
 
-      <_DotnetJSSrcFile Include="$(_WasmRuntimePackSrcDir)\*.js" />
+      <_DotnetJSSrcFile Include="$(_WasmRuntimePackSrcDir)\*.js" Exclude="$(_WasmRuntimePackSrcDir)\*.iffe.js"/>
+      <_DotnetJSPreFile Include="$(_WasmRuntimePackSrcDir)\*.iffe.js"/>
```
Use the existing item for this:
```xml
<_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.iffe.js" Kind="pre-js" />
```

radical commented on src/mono/wasm/runtime/src/mono/init.ts:
```diff
@@ -0,0 +1,498 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Change to use spaces instead of tabs, same as all the other files.

radical commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -337,7 +338,9 @@
 
       <_EmccLinkStepArgs Include="@(_EmccLDFlags)" />
       <_EmccLinkStepArgs Include="--js-library &quot;%(_DotnetJSSrcFile.Identity)&quot;" />
+      <_EmccLinkStepArgs Include="--pre-js &quot;%(_DotnetJSPreFile.Identity)&quot;" />
       <_WasmLinkDependencies Include="@(_DotnetJSSrcFile)" />
+      <_WasmLinkDependencies Include="@(_DotnetJSPreFile)" />
```
And these can be removed:

```suggestion
      <_WasmLinkDependencies Include="@(_DotnetJSSrcFile)" />
```

radical commented:
cc @thays to try this with the debugger tests.

pavelsavara committed:
spaces instead of tabs, draft of lint config

pavelsavara committed:
linker feedback
